toonz/sources/toonz/sceneviewer.cpp

```diff
@@ -1715,17 +1719,15 @@ void SceneViewer::GLInvalidateAll() {
 //-----------------------------------------------------------------------------
 
 void SceneViewer::GLInvalidateRect(const TRectD &rect) {
-  // there is a case that this function is called more than once before
-  // paintGL() is called
-  if (!m_clipRect.isEmpty())
-    m_clipRect += rect;
-  else
+  if (!m_dirty || rect.isEmpty())
     m_clipRect = rect;
```

Comment:
As far as I understand, the purpose of this PR is to separate the following two cases:

- `GLInvalidateRect()` is solely called. (This is normal case)
- `GLInvalidateRect()` is called promptly after `GLInvalidateAll()` before `paintGL()` is executed.

In the former case, Viewer should be updated partially according to the clipping rectangular.
However, in the latter case, whole Viewer should be updated regardless of `GLInvalidateRect()`.
For now `m_clipRect` will be empty for both cases so we need to take measure to distinguish them.

> Or should we change behavior GLInvalidateRect to consider empty rectangle as empty instead of infinite?

Well, yes, in `GLInvalidateAll()` setting `m_clipRect` to `TConsts::infiniteRectD` instead of empty rectangular would be good idea. :+1:

Reply:
> Well, yes, in GLInvalidateAll() setting m_clipRect to TConsts::infiniteRectD instead of empty rectangular would be good idea. +1

It will very good, but i 'm afraid that in this case we shoud to investigate all cases when GLInvalidateRect called from the other code. Program may to call GLInvalidateRect with empty rectangle with meaning "invalidate all".